### ISSUE TITLE
`@stylistic/linebreaks` deletes a bare carriage return instead of writing the break its option asks for

### ISSUE THREAD

**Issue by firefoxic** [stylelint-stylistic/stylelint-stylistic#293]
Found while fixing #270, by asking what the widened write makes reachable. The rule reports a bare carriage return as a line break and then writes over it as though it were a stray character.

### What minimal example or steps are needed to reproduce the bug?

A declaration whose value carries an end-of-line comment closed by a bare carriage return:

```scss
a { b: 1px // c<CR>	2px; }
```

where `<CR>` is a single `\r` and the file ends on a line feed.

The plain-CSS shape of the same write needs no comment at all — two rules parted by a bare carriage return, in a file that ends on a line feed:

```css
a {}<CR>b {}
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "customSyntax": "postcss-scss",
  "plugins": [
    "@stylistic/stylelint-plugin"
  ],
  "rules": {
    "@stylistic/linebreaks": "unix"
  }
}
```

The plain-CSS shape needs the same without `customSyntax`, and reproduces under `postcss-less` as well.

### How did you run Stylelint?

```shell
stylelint "**/*.scss" --config myconfig.json --fix
```

### What did you expect to happen?

The carriage return written as the line feed the option asks for, the way a Windows pair already is.

### What actually happened?

The break is taken away and the two lines are joined. In the first stylesheet that is a file the syntax can no longer read the way it was written: everything behind the double slash is comment text once the break is gone, so the declaration and the block that held it are swallowed, and `--fix` reports a clean pass on the wreck. In the second the two rules are joined into `a {}b {}`, which parses, and no warning survives to say so.

### What is the cause?

`fixData` at [`lib/rules/linebreaks/index.js`](https://github.com/stylelint-stylistic/stylelint-stylistic/blob/main/lib/rules/linebreaks/index.js) opens by deleting every carriage return of the text and then, under `windows`, writes a pair in front of every line feed:

```js
let res = data.replaceAll(`\r`, ``)

if (shouldHaveCR) res = res.replaceAll(`\n`, `\r\n`)
```

That is right for a Windows pair, whose line feed survives the first pass and carries the break, and wrong for a carriage return standing on its own, which has no line feed behind it and is therefore deleted outright. Both options delete it: under `windows` the second pass has nothing left to put a break in front of.

The reading side sees such a break — `hasError` asks `CR_OR_LF` and then `CARRIAGE_RETURN`, so under `unix` a bare carriage return is reported — which is what makes this a member of the class #250 names: one and the same rule counts a character as a line break when it reads and as nothing when it writes.

Two things about the reading are worth naming and are not this issue:

- under `windows` a bare carriage return draws no warning at all, since `hasCR` and `shouldHaveCR` agree, so a file broken with bare carriage returns throughout is passed over;
- the scan splits the file on line feeds alone, so a bare carriage return opens no line of its own and the warning about it is placed at the end of the whole run rather than where the break stands.

### Do you have a proposal to fix the bug?

Respell every break of the set the rule reads, in one pass, rather than deleting one character and writing another:

```js
return data.replaceAll(EVERY_CR_OR_LF_BREAK, shouldHaveCR ? `\r\n` : `\n`)
```

with a name in `lib/regexps.js` matching `\r\n` first so that a Windows pair is one break and never two. The set stays the two characters the rule reads today — a form feed is neither, and widening the rule to it is a reading question that belongs to #250 rather than here.

**Comment by firefoxic** [comment]
The proposal above is not the whole of it, and the branch answering this found the missing half by breaking a file with it.

One text cannot take the break the option asks for. `postcss-less` ends an end-of-line comment on a line feed and on nothing else — `less.parse("// c\r\ta {}")` gives one comment whose text is `c\r\ta {}` — so a carriage return standing in the text of such a comment is text rather than a break of the file. Respelling it ends the comment where the file does not, and everything behind it comes out from under the comment:

```less
a { color: red; }
// off: d {<CR>color: blue; }
```

Deleting the carriage return, which is what the fix does today, leaves that comment closed and the code under it. Respelling it as a line feed leaves a file `postcss-less` cannot read at all, `Unexpected }` at the closing brace, and the same write turns commented-out code live wherever the comment happens to close a balanced block:

```less
a { color: red; }
// b { color: blue; }<CR>c { color: green; }
```

comes back with `c` as a live rule.

So the write is respelled everywhere but there. In the text of a comment the syntax closes with a line break, the carriage return is taken out instead — what the fix has always done to every one of them, and the right thing to do to this one. `writesIntoInlineComment` is what puts the question; the same name #211 left for the other rules that write beside such a comment.

Nothing here reaches `postcss-scss`, which ends an end-of-line comment on a carriage return as readily as on a line feed, so no break can stand in such a comment's text under that syntax at all.

**Comment by firefoxic** [comment]
The comment above is wrong, and the branch does not implement what it proposes. Keeping it here rather than editing it, since it is what led to the measurement that overturned it.

`less.parse` in that comment is `postcss-less`, not Less. The Less compiler normalises the line endings of a file before it parses — `lib/less/parser/parser.js`, `str = str.replace(/\r\n?/g, '\n')`, unconditional, right before the input is started — so a bare carriage return closes an end-of-line comment in Less exactly as a line feed does. It is `postcss-less` that disagrees with the compiler it models, its `isInlineComment` reading `/\r?\n/`. dart-sass ends such a comment on a carriage return too, and on a form feed besides; `postcss` tokenises a carriage return as whitespace; and CSS Syntax §3.3 maps a carriage return, a form feed and a Windows pair alike onto a line feed before anything else happens. Five places in `lib/` already said as much — `endsWithInlineComment`, `endsInlineCommentOnFormFeed`, `EVERY_FORM_FEED` and the two `function-parentheses-*-inside` guards.

Compiled with the real Less compiler, the two examples above say the opposite of what the comment claims:

| source | Less reads | after the fix as proposed above | after the fix as it ships |
| --- | --- | --- | --- |
| `a{color:red}<CR>// b{color:blue}<CR>c{color:green}` | `a` and `c` are live | only `a` — `c` is commented out | `a` and `c` |
| `// one<CR>a{b:c}<CR>// two<CR>d{e:f}` | two live rules | the whole file commented out | two live rules |

So the carve-out was not a guard against corruption; it was the corruption, on any Less file written with bare carriage returns throughout — which is the file this rule exists to repair. The text of an end-of-line comment is respelled like every other text.

One thing does follow from the parser's disagreement, and it is worth naming rather than hiding. `a { color: red; }<CR>// off: d {<CR>color: blue; }` is a Less parse error before the fix and after it — `Unrecognised input. Possibly missing opening '{'` — but `postcss-less` accepts it before and rejects it after, because it alone was still holding the second line inside the comment. A file that lints today and errors tomorrow is a real annoyance; it is an already-broken file either way, and the alternative was the two rows above.
